### PR TITLE
feat(pubsub): reduce contention in lease management

### PR DESCRIPTION
**Pull request by coryan**
I found that the lease management layer suffered from too much
contention. Basically each `ack()` and `nack()` call required a lock, as
the hash map with all the pending messages is updated. After updating
this hash map the code calls the `StreamingSubscriptionBatchSource`
stage to do more work, which requires additional locking. This PR
collapses the functionality of the `SubscriptionLeaseManagement` stage
onto `StreamingSubscriptionBatchSource`, reducing the number of locks
per `ack()`/`nack()` call.

These improvements in locking overhead are minor, but collapsing the two
layers simplifies other changes I want to make, such as (1) suspend
`Read()` calls if the number (and/or size) of outstanding messages is
too high, (2) create multiple AsyncPull streams on the same session, and
(3) maybe avoid copying the large `ack_id` strings (160 bytes!) and use
a more efficient in-memory representation.

I also simplified the timer management to periodically update the
leases. We use a single periodic timer to check if the leases need an
update, and piggy back the update with any other pending nacks and/or
acks. That is less accurate, but the periodic timer is executed about
100 times a second, and the leases need to be updated every 10 seconds
(we actually do it ever 5). Good enough I think.

Part of the work for #5512 and #5431 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5549)

### CODE REVIEW

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.cc [hunk 1]
```diff
@@ -316,34 +322,52 @@ void StreamingSubscriptionBatchSource::OnFinish(Status status) {
...
   acks.swap(ack_queue_);
   std::vector<std::string> nacks;
   nacks.swap(nack_queue_);
-  std::vector<std::pair<std::string, std::chrono::seconds>> deadlines;
-  deadlines.swap(deadlines_queue_);
+
+  std::vector<std::string> lease_extensions;
+  // We try to avoid work while holding the `lk` lock, but rarely (about once
+  // every 5 seconds) we need to update the leases. That requires a potentially
+  // long loop.
```
Comment (devbww) [inline google/cloud/pubsub/internal/streaming_subscription_batch_source.cc:347]: s/loop/time/ perhaps?

Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.cc]: Fixed. And fixed other things as I was looking at this code again.

File: google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc [hunk 2]
```diff
@@ -658,16 +645,36 @@ TEST(StreamingSubscriptionBatchSourceTest, AckBatching) {
   auto subscription = pubsub::Subscription("test-project", "test-subscription");
   std::string const client_id = "fake-client-id";
 
+  // We will configure StreamingSubscriptionBatchSource's to use this period
+  // in its periodic timer.
+  auto constexpr kHz = std::chrono::milliseconds(42);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source_test.cc]: Fixed.

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.h [hunk 3]
```diff
@@ -59,18 +61,58 @@ struct AckBatchingConfig {
...
+
+  static auto constexpr kMinimumAckDeadline = std::chrono::seconds(10);
+  static auto constexpr kMaximumAckDeadline = std::chrono::seconds(600);
+
+  struct TimerConfig {
+    TimerConfig() : clock([] { return std::chrono::system_clock::now(); }) {}
+
+    Clock clock;
+    std::chrono::milliseconds period{10};
+    std::chrono::milliseconds lease_refresh{kMinimumAckDeadline / 2};
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.h:123]: Yes. I am not sure it matters.  We need something that happens more often than the minimum period, but could be 1 or 2 seconds less or 1/2.  I guess "1 second" is a problem, this is unlikely to happen as the minimum is set by the service.  Added some comments to explain where these are coming from.

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.h [hunk 4]
```diff
@@ -59,18 +61,58 @@ struct AckBatchingConfig {
...
+ * - Once successfully connected [1], it resumes the request if the streaming
+ *   pull is terminated with any failure.
+ * - Sends at most one `Read()` and at most one `Write()` request to the
+ *   underlying gRPC abstraction.
+ * - On shutdown, wait for any pending `Read()` and `Write()` requests to
+ *   complete before calling `Finish()`.
+ * - Invoke a callback when `Read()` receives data.
+ * - To preserve ordering, only once the callback returns is a new `Read()`
+ *   request issued.
+ * - To minimize I/O buffers nacks and acks into a single `Write()` request.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.h]: Done.

File: google/cloud/pubsub/internal/streaming_subscription_batch_source.h [hunk 5]
```diff
@@ -59,18 +61,58 @@ struct AckBatchingConfig {
   std::chrono::milliseconds max_hold_time{100};
 };
 
+/**
+ * Keep a single StreamingPull active.
+ *
+ * Creates a single StreamingPull request active:
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/internal/streaming_subscription_batch_source.h]: Yeah, fixed.

**Comment by coryan** [comment]
PTAL

**Review by devbww — APPROVED** [review]

**Comment by coryan** [comment]
I am going to close this for now, I noticed some regressions (increased flakiness) in local testing and cannot seem to be able to fix them. Sorry to waste your time @devbww.

**Pull request closed without merging**